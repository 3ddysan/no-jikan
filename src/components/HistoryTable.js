import React from 'react'
import { library } from '@fortawesome/fontawesome-svg-core'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faCalendarDay, faClock, faCoffee } from '@fortawesome/free-solid-svg-icons'

library.add(faCalendarDay, faClock, faCoffee)

function list(entries, onDelete) {
    if (entries.length > 0) {
        return entries.map((entry, index) => (
            <tr key={index} className="on-mobile-higher">
                <td>{entry.work}</td>
                <td>{entry.break}</td>
                <td className="has-text-centered">{entry.date}</td>
                <td className="is-hidden-mobile has-text-centered">
                    <button className="button is-light" onClick={() => onDelete(index)}>
                        <span className="">Delete</span>
                    </button>
                </td>
            </tr>
        ))
    } else {
        return (
            <tr>
                <td colSpan="4" className='has-text-centered'>
                    There's no data
                </td>
            </tr>
        );
    }
}

const HistoryTable = ({ entries = [], onDelete }) => (
    <div className="container">
        <table className="table is-fullwidth">
            <thead>
                <tr>
                    <th>
                        <FontAwesomeIcon icon="clock" />
                        <span className="is-hidden-mobile"> Time</span>
                    </th>
                    <th>
                        <FontAwesomeIcon icon="coffee" />
                        <span className="is-hidden-mobile"> +Break</span>
                    </th>
                    <th className="has-text-centered">
                        <FontAwesomeIcon icon="calendar-day" />
                        <span className="is-hidden-mobile"> Date</span>
                    </th>
                    <th className="has-text-centered is-hidden-mobile">
                        <span>Actions</span>
                    </th>
                </tr>
            </thead>
            <tbody>
                {list(entries, onDelete)}
            </tbody>
        </table>
    </div>
);

export default HistoryTable;