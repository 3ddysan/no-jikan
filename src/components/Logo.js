import React from 'react'
import styled from 'styled-components'

const Title = styled.h1`
    width: 75px;
    background-color: lightcoral;
    padding-left: 2px;
    box-shadow: 0px 0px 1px 5px rgba(0,0,0,1);

    span {
        background-color: white;
    }

    transition: all 0.3s ease-in-out 0s;

    &:hover {
        cursor: pointer;
        transform: rotate(8deg);
        transition: all 0.3s ease-in-out 0s;
    }
`

const Logo = () => (
    <div className="container">
        <Title className="title">no<span>-</span>jikan</Title>
    </div>
);

export default Logo;