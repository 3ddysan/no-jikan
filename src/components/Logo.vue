<template>
  <div class="logo">
    <h1 class="title">no<span>-</span>jikan</h1>
  </div>
</template>
<style scoped>
.logo {
  margin: 1em;
}
.title {
  width: 70px;
  background-color: lightcoral;
  padding-left: 2px;
  box-shadow: 0px 0px 1px 5px hwb(0 0% 100%);
  font-weight: bold;
}

.title span {
  background-color: white;
}

.title:hover {
  cursor: pointer;
  transform: rotate(8deg);
  transition: all 0.3s ease-in-out 0s;
}
</style>
