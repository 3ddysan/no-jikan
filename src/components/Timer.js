import React, { useEffect, useState, useRef } from 'react'
import styled from 'styled-components'
import { classes } from 'helper'

const TimerDisplay = styled.div`
padding-top: 15px;
padding-bottom: 15px;
`

const TimeUnit = styled.div`
padding: 10px;
background: #F9F9F9;
display: inline-block;

.unit {
    padding: 15px;
    border-radius: 3px;
    background: #F0F0F0;
    display: inline-block;
    width: 60px;
}

.unit-label {
    padding-top: 3px;
}
`
const pad = (v) => v.toString().padStart(2, 0)

const formatSeconds = (s) => {
    const minutes = pad(Math.floor(s / 60) % 60);
    const hours = pad(Math.floor(s / 3600));
    const seconds = pad(s % 60);
    return { hours, minutes, seconds };
}

export const useTimer = (active = false, elapsedSeconds = 0) => {
    const [state, setState] = useState('stopped')
    const [seconds, setSeconds] = useState(elapsedSeconds)
    const [isActive, setActive] = useState(active)
    const intervalRef = useRef()

    useEffect(() => {
        if (isActive) {
            setState('running');
            intervalRef.current = setInterval(() => {
                setSeconds(previous => previous + 1)
            }, 1000);
            return () => {
                clearInterval(intervalRef.current)
            }
        }
        clearInterval(intervalRef.current)
        setState(s => s === 'running' ? 'paused' : 'stopped');
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [isActive])

    const toggle = () => {
        setActive(isActive => !isActive)
    }

    const reset = () => {
        setActive(false)
        setSeconds(0)
        setState('stopped');
    }

    return {
        timer: formatSeconds(seconds),
        toggle,
        reset,
        state
    };
}

const playPauseStyle = {
    stopped: 'is-primary',
    paused: 'is-warning',
    running: 'is-active',
};

const Timer = ({ active = false, elapsedSeconds = 0 }) => {
    const { timer, toggle, reset, state } = useTimer(active, elapsedSeconds);
    const { hours, minutes, seconds } = timer;
    return (
        <>
            <div className="container">
                <TimerDisplay>
                    <TimeUnit>
                        <div className="unit is-size-5">{hours}</div>
                        <div className="unit-label is-size-7 has-text-centered">Hours</div>
                    </TimeUnit>
                    <TimeUnit>
                        <div className="unit is-size-5">{minutes}</div>
                        <div className="unit-label is-size-7 has-text-centered">Minutes</div>
                    </TimeUnit>
                    <TimeUnit>
                        <div className="unit is-size-5">{seconds}</div>
                        <div className="unit-label is-size-7 has-text-centered">Seconds</div>
                    </TimeUnit>
                </TimerDisplay>
            </div>
            <div className="container">
                <div className="columns is-mobile is-centered">
                    <div className="column is-1" style={{ width: 'auto' }}>
                        <button onClick={toggle} className={classes('button', 'is-medium', playPauseStyle[state])}>
                            {state === 'running' ? 'Pause' : 'Play'}
                        </button>
                    </div>
                    <div className="column is-1" style={{ width: 'auto' }}>
                        <button onClick={reset} className="button is-danger is-medium" disabled={state === 'stopped'}>
                            Stop
                        </button>
                    </div>
                </div>
            </div>
        </>
    )
};

export default Timer;