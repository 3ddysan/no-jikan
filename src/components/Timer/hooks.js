import { useEffect, useState, useRef } from 'react'

const formatSeconds = (s) => {
    const minutes = Math.floor(s / 60) % 60;
    const hours = Math.floor(s / 3600);
    const seconds = s % 60;
    return { hours, minutes, seconds };
}

export const useTimer = (active = false, elapsedSeconds = 0) => {
    const [seconds, setSeconds] = useState(elapsedSeconds)
    const [isActive, setActive] = useState(active)
    const intervalRef = useRef()
    const state = isActive ? 'running' : seconds > 0 ? 'paused' : 'stopped'

    useEffect(() => {
        if (isActive) {
            intervalRef.current = setInterval(() => {
                if (state === 'running') {
                    setSeconds(previous => previous + 1)
                }
            }, 1000);
            return () => {
                clearInterval(intervalRef.current)
            }
        }
        clearInterval(intervalRef.current)
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [isActive])

    const toggle = () => {
        setActive(isActive => !isActive)
    }

    const reset = () => {
        setActive(false)
        setSeconds(0)
    }

    return {
        timer: formatSeconds(seconds),
        toggle,
        reset,
        state
    };
}
