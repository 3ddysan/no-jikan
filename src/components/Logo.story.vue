<script setup lang="ts">
import Logo from './Logo.vue';
</script>
<template>
  <Story>
    <Logo />
  </Story>
</template>
