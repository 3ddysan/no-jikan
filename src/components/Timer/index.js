import React from 'react'
import styled from 'styled-components'
import { classes } from 'helper'
import { useTimer } from 'components/Timer/hooks'

const TimerDisplay = styled.div`
    padding-top: 15px;
    padding-bottom: 15px;
`

const TimeUnit = styled.div`
    padding: 10px;
    display: inline-block;
    ${({ paused }) => paused ? `
        background: #fff3c2;
    ` : `
        background: #F9F9F9;
    `}

    .unit {
        padding: 15px;
        border-radius: 3px;
        display: inline-block;
        width: 60px;
        ${({ paused }) => paused ? `
            background: #ffe26c;
        ` : `
            background: #F0F0F0;
        `}
    }

    .unit-label {
        padding-top: 3px;
    }
`

const pad = (v) => v.toString().padStart(2, 0)

export function Display({ hours, minutes, seconds, paused }) {
    return (
        <TimerDisplay paused={paused} data-testid="display">
            <TimeUnit paused={paused}>
                <div className="unit is-size-5">{pad(hours)}</div>
                <div className="unit-label is-size-7 has-text-centered">Hours</div>
            </TimeUnit>
            <TimeUnit paused={paused}>
                <div className="unit is-size-5">{pad(minutes)}</div>
                <div className="unit-label is-size-7 has-text-centered">Minutes</div>
            </TimeUnit>
            <TimeUnit paused={paused}>
                <div className="unit is-size-5">{pad(seconds)}</div>
                <div className="unit-label is-size-7 has-text-centered">Seconds</div>
            </TimeUnit>
        </TimerDisplay>
    );
}

const playPauseStyle = {
    stopped: 'is-primary',
    paused: 'is-warning',
    running: 'is-active',
};

export function Controls({ onToggle, onStop, state }) {
    return (
        <div className="columns is-mobile is-centered">
            <div className="column is-1" style={{ width: 'auto' }}>
                <button data-testid="startButton" onClick={onToggle} className={classes('button', 'is-medium', playPauseStyle[state])}>
                    {state === 'running' ? 'Pause' : 'Start'}
                </button>
            </div>
            <div className="column is-1" style={{ width: 'auto' }}>
                <button data-testid="stopButton" onClick={onStop} className="button is-danger is-medium" disabled={state === 'stopped'}>
                    Stop
            </button>
            </div>
        </div>
    );
}

const Timer = ({ active = false, elapsedSeconds = 0, onFinish }) => {
    const { timer, toggle, stop, state, totalSeconds } = useTimer(active, elapsedSeconds);
    const onStop = () => {
        onFinish && onFinish(totalSeconds);
        stop();
    };
    return (
        <>
            <div className="container">
                <Display {...timer} />
            </div>
            <div className="container">
                <Controls onToggle={toggle} onStop={onStop} state={state} />
            </div>
        </>
    )
};

export default Timer;