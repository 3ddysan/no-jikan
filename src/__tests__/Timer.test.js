import React from 'react';
import { Controls, Display } from 'components/Timer';
import renderer from 'react-test-renderer';
import { render, fireEvent, cleanup } from 'react-testing-library';

describe('Display', () => {

  it('renders state correctly', () => {
    const tree = renderer
      .create(<Display hours="02" minutes="41" seconds="03" />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

});

describe('Controls', () => {

  it('renders paused state correctly', () => {
    const tree = renderer
      .create(<Controls state='paused' />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders running state correctly', () => {
    const tree = renderer
      .create(<Controls state='running' />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders stopped state correctly', () => {
    const tree = renderer
      .create(<Controls state='stopped' />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  afterEach(cleanup);

  it('should toggle play/pause callback', () => {
    const toggle = jest.fn();
    const reset = jest.fn();
    const { getByText } = render(<Controls onToggle={toggle} onReset={reset} state='stopped' />);

    fireEvent.click(getByText('Play'));

    expect(toggle).toBeCalled();
    expect(reset).not.toBeCalled();
  });

  it('should toggle reset callback', () => {
    const toggle = jest.fn();
    const reset = jest.fn(() => console.log('click on Stop button'));
    const { getByText } = render(<Controls onToggle={toggle} onReset={reset} state='running' />);

    fireEvent.click(getByText('Stop'));

    expect(reset).toBeCalled();
    expect(toggle).not.toBeCalled();
  });

});