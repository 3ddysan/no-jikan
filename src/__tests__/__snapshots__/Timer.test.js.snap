// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Controls renders paused state correctly 1`] = `
<div
  className="columns is-mobile is-centered"
>
  <div
    className="column is-1"
    style={
      Object {
        "width": "auto",
      }
    }
  >
    <button
      className="button is-medium is-warning"
    >
      Play
    </button>
  </div>
  <div
    className="column is-1"
    style={
      Object {
        "width": "auto",
      }
    }
  >
    <button
      className="button is-danger is-medium"
      disabled={false}
    >
      Stop
    </button>
  </div>
</div>
`;

exports[`Controls renders running state correctly 1`] = `
<div
  className="columns is-mobile is-centered"
>
  <div
    className="column is-1"
    style={
      Object {
        "width": "auto",
      }
    }
  >
    <button
      className="button is-medium is-active"
    >
      Pause
    </button>
  </div>
  <div
    className="column is-1"
    style={
      Object {
        "width": "auto",
      }
    }
  >
    <button
      className="button is-danger is-medium"
      disabled={false}
    >
      Stop
    </button>
  </div>
</div>
`;

exports[`Controls renders stopped state correctly 1`] = `
<div
  className="columns is-mobile is-centered"
>
  <div
    className="column is-1"
    style={
      Object {
        "width": "auto",
      }
    }
  >
    <button
      className="button is-medium is-primary"
    >
      Play
    </button>
  </div>
  <div
    className="column is-1"
    style={
      Object {
        "width": "auto",
      }
    }
  >
    <button
      className="button is-danger is-medium"
      disabled={true}
    >
      Stop
    </button>
  </div>
</div>
`;

exports[`Display renders state correctly 1`] = `
<div
  className="sc-bdVaJa gZRiZe"
>
  <div
    className="sc-bwzfXH gCXOUF"
  >
    <div
      className="unit is-size-5"
    >
      02
    </div>
    <div
      className="unit-label is-size-7 has-text-centered"
    >
      Hours
    </div>
  </div>
  <div
    className="sc-bwzfXH gCXOUF"
  >
    <div
      className="unit is-size-5"
    >
      41
    </div>
    <div
      className="unit-label is-size-7 has-text-centered"
    >
      Minutes
    </div>
  </div>
  <div
    className="sc-bwzfXH gCXOUF"
  >
    <div
      className="unit is-size-5"
    >
      03
    </div>
    <div
      className="unit-label is-size-7 has-text-centered"
    >
      Seconds
    </div>
  </div>
</div>
`;
